### PR TITLE
Fix admin GUI edit for users with no subscription record

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- The admin Account Manager could only *update* an existing subscription. For a
  user with no `subscription_users` row (never-subscribed free user, or a user
  whose subscription was removed during test cleanup), `PUT
  /admin/users/{id}/subscription` returned `400 "User has no subscription
  record"` and there was no create path — the only recourse was raw SQL.
- `update_user_subscription_admin` is now an **upsert**: when the
  `subscription_users` and/or `user_storage_usage` row is missing it is created
  with the submitted plan/quota; when it exists the prior behavior is unchanged.
  The audit log records missing rows as `null` in `old_value` so the entry
  reads as "created" rather than "edited".

#### Design Decisions
- **Upsert the existing PUT rather than add a POST.** Smallest change that makes
  the one endpoint the UI already calls complete; no new route, no frontend
  branching on "does a record exist".
- **Audit snapshot fields made optional.** `SubscriptionAuditSnapshot.plan_id`
  and `storage_quota_bytes` were non-nullable; recording a genuinely-absent row
  as `None` would have raised a pydantic `ValidationError` and turned the 400
  into a 500. They are now `Optional`. `new_value` is still always fully
  populated from `UserSubscriptionUpdate`, so the relaxation only affects
  `old_value`.
- **Two independent existence flags, not one.** The two rows are queried
  separately and can be missing independently (partial cleanup). Each of
  `subscription_existed` / `storage_existed` gates only its own field in the
  audit `old_value`, so a mixed state is recorded truthfully.
- **Concurrency guard mirrors the checkout path.** Both target tables are unique
  on `user_id`. `_insert_or_reselect` wraps each insert in a `SAVEPOINT`
  (`begin_nested`) and, on `IntegrityError`, re-selects the row a concurrent
  writer created — the same pattern `CheckoutService` already uses — instead of
  surfacing a spurious 500.
- **Premium requires an expiration.** A new premium subscription created without
  an expiration would be born already expired (`expiration or now()`). The
  endpoint now rejects premium with no expiration (400); free still defaults to
  now, which is meaningless for free and matches signup provisioning.
- **Storage bucket intentionally not provisioned here.** This is a
  record-repair path; a user missing these rows in practice already has a bucket
  from signup. Documented inline with the upgrade path
  (`ensure_user_bucket_exists`) if that assumption ever changes. Bucket/storage
  provisioning is out of scope per the issue.

### Evidence
```
$ docker-compose -f docker-compose.test.yml run --rm -w /app test_studio_backend \
    python -m pytest studio/tests/app/common/routers/test_users_admin_subscription.py -q
20 passed, 43 warnings in 0.38s

# no cross-contamination from the test module's sqlite type/DDL shims:
$ ... python -m pytest studio/tests/app/common/routers \
    studio/tests/app/common/core/subscription studio/tests/app/common/schemas -q
495 passed, 3 skipped, 45 warnings in 16.99s
```

### References
- Issue #720 — "Admin GUI cannot change a user's plan / quota if no subscription record"
- Prior art for the concurrency guard: `CheckoutService` subscription insert
  (`begin_nested` + `IntegrityError` fallback-to-update).

#### Files changed
- `studio/app/common/core/users/crud_users.py` — `update_user_subscription_admin`
  now upserts both rows instead of 400-ing when they are missing; independent
  `subscription_existed` / `storage_existed` flags drive an honest audit
  `old_value`; premium without an expiration is rejected with 400. New helper
  `_insert_or_reselect` guards the inserts against the unique-constraint race.
- `studio/app/common/schemas/users.py` — `SubscriptionAuditSnapshot.plan_id` and
  `storage_quota_bytes` made `Optional[int]` so `old_value` can represent a row
  that did not exist.
- `studio/tests/app/common/routers/test_users_admin_subscription.py` — added a
  DB-backed `TestUpdateUserSubscriptionAdminCRUD` class exercising the upsert,
  plus snapshot tests for the all-null / mixed-null cases. Includes SQLite shims
  (BIGINT→INTEGER compilation, stripping MySQL `ON UPDATE` defaults, per-table
  `create_all`) so the MySQL-shaped models can run under an in-memory session.

### Manual Testcases
1. **Never-subscribed user.** As an org admin, pick a user where `SELECT * FROM
   subscription_users WHERE user_id = {N}` is empty. In the Account Manager set
   their plan and storage quota and save.
   - Expected: request returns 200; `subscription_users` and `user_storage_usage`
     rows now exist with the chosen plan and `storage_quota_bytes`; a
     `subscription_audit_log` row records the change with `old_value` fields
     null.
2. **Existing subscriber (regression).** As an admin, change plan/quota for a
   user who already has both rows.
   - Expected: 200; rows updated in place; audit `old_value` reflects the real
     prior plan/quota.
3. **Premium without expiration.** As an admin, submit `plan_id = premium` with
   no expiration.
   - Expected: 400 "expiration is required for the premium plan"; no rows
     created or changed.
4. Automated: `docker-compose -f docker-compose.test.yml run --rm -w /app
   test_studio_backend python -m pytest
   studio/tests/app/common/routers/test_users_admin_subscription.py` — all 20
   tests pass.

### Unit, Integration, Contract Test Coverage
- Schema: existing `UserSubscriptionUpdate` / `SubscriptionAuditSnapshot`
  validation tests, plus new all-null and mixed-null snapshot cases.
- CRUD (in-memory DB session): both rows created when missing; subscription
  created when only storage exists (mixed, audit records real old quota + null
  plan); existing rows updated with prior values captured in the audit; premium
  without expiration rejected and nothing written.

#### Difficulties
- No DB-session fixture existed for these CRUD functions. The models use
  MySQL-specific constructs (`BIGINT` primary keys, `ON UPDATE CURRENT_TIMESTAMP`
  server defaults) that in-memory SQLite cannot create or autoincrement. The
  test builds only the four tables it touches, strips the MySQL `ON UPDATE`
  default, and compiles `BIGINT`→`INTEGER` on the SQLite dialect. These shims
  are confined to the test module and were verified not to affect the broader
  suite (495 passed).

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| `update_user_subscription_admin` (admin-only) | Low | Single admin endpoint; existing update path behavior unchanged; new branch covered by tests. |
| `SubscriptionAuditSnapshot` schema | Low | Only consumers are this function's two `.dict()` calls into JSON audit columns; nothing reads the fields back expecting non-null; `new_value` still fully populated. |
| Concurrency | Low | `begin_nested` + `IntegrityError` fallback mirrors the proven checkout path; on the near-impossible double-admin race one request updates instead of 500-ing. |
| Storage bucket | Low | Path intentionally does not provision a bucket; target users already have one from signup. Documented inline. Separate from the raw-vs-effective quota enforcement bug tracked in the issue. |

### CODE REVIEW

File: studio/tests/app/common/routers/test_users_admin_subscription.py [hunk 1]
```diff
@@ -174,3 +201,209 @@ def test_snapshot_dict_serialization(self):
...
+    ]
+    # Drop the MySQL "ON UPDATE CURRENT_TIMESTAMP" default (invalid SQLite DDL).
+    for table in tables:
+        for col in table.columns:
+            arg = getattr(col.server_default, "arg", None)
+            if arg is not None and "ON UPDATE" in str(arg):
+                col.server_default = None
+    SQLModel.metadata.create_all(engine, tables=tables)
+    with Session(engine) as session:
+        yield session
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/tests/app/common/routers/test_users_admin_subscription.py]: Addressed in e0d0ae0c9

**Review by itutu-tienday — COMMENTED** [review]
**Overall verdict: Approve with minor changes.** The change is well-scoped, the design choices are sound and clearly documented, and it reuses an existing, proven concurrency pattern. Two themes drive the findings below: (1) the new concurrency guard has no test coverage, and (2) the endpoint now *implicitly materializes records* (an upsert), which is largely safe but leaves a user in a **partially-provisioned** state that only *looks* complete — the sharpest residual concern.

---

## 1. Design & Specification Validity

| Decision | Assessment |
|----------|-----------|
| Upsert the existing `PUT /admin/users/{id}/subscription` instead of adding a `POST` | **Good.** Smallest change that completes the endpoint the UI already calls. Router (`users_admin.py:78`) is a thin delegate with no duplicate guard, so no router change was needed — correct. |
| Two independent existence flags (`subscription_existed` / `storage_existed`) | **Good.** The two rows can be missing independently (partial cleanup). Gating each audit field separately produces an honest `old_value` for mixed states. |
| Relaxing `SubscriptionAuditSnapshot.plan_id` / `storage_quota_bytes` to `Optional` | **Safe.** Only two consumers exist (`crud_users.py:606` `old_value`, `:621` `new_value`), both in this function. `new_value` is always fully populated, so the relaxation only affects `old_value`. Nothing reads these fields back expecting non-null. |
| Concurrency guard mirrors `CheckoutService` (`begin_nested` + `IntegrityError` → re-select) | **Good.** Both `subscription_users` and `user_storage_usage` are unique on `user_id`; identical to `checkout_service.py:381` / `stripe_service.py:69`. Reusing a proven idiom is the right call. |
| Premium requires an explicit expiration | **Correct fix, but a behavior change** — see §4.4. |
| Storage bucket intentionally not provisioned | **The crux of the implicit-creation concern** — see §3. |

The core spec — "make the endpoint create the missing rows instead of 400-ing, and record the creation truthfully in the audit log" — is met.

---

## 2. Is the dynamic (implicit) record creation itself harmful?

The endpoint now creates `subscription_users` / `user_storage_usage` rows on demand. This was assessed against how the two rows are created and consumed elsewhere. **The mechanism itself is largely safe** — for the two rows it touches, the admin upsert builds them almost identically to the canonical signup provisioning (`create_user`, `crud_users.py:413-427`). The concerns that turned out **not** to be problems:

| Potential harm | Finding |
|----------------|---------|
| Free row born with `expiration = now()` (immediately "expired") gets swept by an expiration job | **Not triggered.** Both `get_users_for_expiration_deletion` (`subscription_service.py:494`) and `premium_expiration_sweep_job` (`:152`) filter `plan_id == PREMIUM`. A free row with a past expiration is never swept or deleted. Canonical signup does the same `expiration = now()` for free — behavior is consistent. |
| `sync_status = pending` on the created row causes a background Stripe-sync job to fire / create a phantom subscription | **Not triggered.** `sync_status` is only *written* (SYNCED/FAILED) reactively; no job *reads* `pending` as a driver. Canonical signup also leaves it `pending`. The admin path bypasses Stripe by design, same as the existing update path. |
| Concurrent double-create | Guarded by the unique constraint + `_insert_or_reselect` SAVEPOINT fallback. |
| Premium row born already expired | Prevented by the new expiration-required check (§4.4). |

So the residual harms are **not** about the act of creating rows; they are about creating *only a subset* of what a fully-provisioned user has, and about masking why the rows were missing. Those are §3 and §4.

---

## 3. [Medium] Partial provisioning — the upsert restores rows but not the full signup invariant

The canonical creation path `create_user` (`crud_users.py:401-427`) materializes the two rows **and** an S3 bucket via `ensure_user_bucket_exists`. The admin upsert deliberately creates only the two rows and skips the bucket (documented inline). It can therefore produce a user with a `subscription_users` row but **no bucket** — a state that looks complete to the admin UI but is not.

- The PR's justification ("target users already have a bucket from signup") holds for the **test-cleanup case** (rows deleted, bucket intact) but **not** for a genuinely never-provisioned user. If such a user is granted premium and then uploads, the flow will 500 unless the upload/runtime path itself calls `ensure_user_bucket_exists`.
  - **Action:** confirm the upload / premium-activation runtime path lazily ensures the bucket. If it does, this is fully benign and worth a one-line note in the PR; if it does not, this admin path should ensure the bucket too.
- Related divergences from canonical provisioning:
  - **Free-plan quota is not the canonical value.** Signup uses `StorageQuota.FREE * StorageSize.GB`; the upsert stores the admin-submitted `data.storage_quota_bytes`. A free user can end up with a non-standard quota. Admin discretion, but inconsistent — consider using the canonical FREE constant when `plan_id == FREE`. (§4.5)
  - `FreeUserAssignment` is created lazily by `user_activity_middleware.py:441`, so its absence here is normal and self-healing — **not** a concern.

**Bottom line:** implicit creation is safe *per row*, but the endpoint reconstructs a subset of the provisioning invariant. Name this explicitly and confirm the bucket assumption.

---

## 4. Findings

### 4.1 [Medium — Test coverage] The concurrency guard `_insert_or_reselect` has zero test coverage
`_insert_or_reselect` (`crud_users.py:552`) — the SAVEPOINT + `IntegrityError` → re-select fallback — is the highest-risk new code and **no test exercises the fallback branch.** All four new CRUD tests hit only the happy insert path. Untested: the `except IntegrityError:` re-select, and the `if existing is None: raise` re-raise. The race is hard to reproduce under SQLite, but the fallback is unit-testable without a real race (pre-insert a conflicting row and force `flush()` to raise `IntegrityError`; assert the existing row is returned; a second case asserting re-raise when the row can't be found).

### 4.2 [Medium — Observability] Implicit repair masks the root cause; no operational signal
Auto-creating on admin edit silently repairs a state that "should never happen." If rows go missing due to a real bug (not just test cleanup), the upsert hides the symptom — the previous 400 was at least a signal. The audit log records `old_value = null` (a trace exists), but it is a passive audit-table record, not an alert. **Action:** emit a `WARNING` log when the admin path has to *create* (rather than update) a row, so a systemic cause of missing rows is not masked.

### 4.3 [Low–Med — Global side effect mislabeled] `@compiles(..., "sqlite")` is process-global, not module-scoped
```python
@compiles(BigInteger, "sqlite")
@compiles(GENERIC_BIGINT, "sqlite")
@compiles(MYSQL_BIGINT, "sqlite")
def _bigint_as_integer_sqlite(...): return "INTEGER"
```
The PR/comment says the shims are "confined to the test module." `@compiles` mutates SQLAlchemy's **global** dialect-compiler registry the moment the module is imported; any *other* SQLite-backed test in the same process that builds DDL from a `BigInteger`/`BIGINT` column will now emit `INTEGER`. The "495 passed" run shows no current conflict, but this is latent cross-test coupling. **Action:** correct the wording, or scope the coercion to a private `TypeDecorator`/dedicated engine rather than a global `@compiles`. The fixture also mutates process-global `Table.column.server_default` in place (restored in `finally`) — correct for serial runs; worth a comment acknowledging the global nature.

### 4.4 [Low — Spec/behavior change] Premium-without-expiration now 400s on the *update* path too
The new guard rejects `plan_id = premium` with `expiration is None` for **both** create and the pre-existing update path. Previously an update to premium with no expiration silently coerced `expiration = now()` (born-expired — a real bug). The fix is correct, but it turns a previously-succeeding (if buggy) admin update into a hard 400. Safe only if the Account Manager UI always submits an expiration for premium — please confirm no frontend flow relied on the old silent coercion.

### 4.5 [Low — Data correctness] Created storage row uses `storage_usage_bytes=0` and non-canonical quota
For a user who genuinely has stored data but lost the row, the repair inserts `storage_usage_bytes = 0`, under-reporting real usage until the next full S3 scan reconciles it (`delta_since_last_scan` / `last_full_scan`). Confirm a scan reconciles the count. Quota uses the admin-submitted value rather than the canonical FREE constant for free plans (see §3).

### 4.6 [Nit] Minor code-quality items
- `_insert_or_reselect(db, row, model, user_id)` — `row`/`model` untyped; a couple of hints or a docstring line naming the expected models would aid readability.
- Free-plan creation writes `expiration = now()` into the audit `new_value` (meaningless for free). Pre-existing and consistent — noted only because it now persists into newly-created rows.

---

## 5. Code Quality
- Clear inline comments explaining *why* (bucket omission, audit-null rationale, savepoint race) rather than *what*. Matches repo conventions.
- Create-then-assign avoids duplicating field assignments across create/update branches.
- Reuses the established `begin_nested` + `IntegrityError` idiom instead of inventing one.
- Error handling: outer `try/except HTTPException / except Exception` rolls back and maps unexpected errors to 500; the re-select-None `raise` is properly surfaced through it.

---

## 6. Test Sufficiency
**Present and good:** schema all-null / mixed-null snapshots; CRUD (real in-memory session) for both rows created, subscription-created-when-only-storage-exists (mixed audit), existing rows updated with prior values captured, premium-without-expiration rejected with nothing written.

**Gaps:**
1. `_insert_or_reselect` `IntegrityError` fallback and its `None` re-raise are untested (§4.1) — highest-risk new code.
2. Only one of the two mixed states is covered at CRUD level; the inverse ("subscription exists, storage missing") would fully lock down the independent-flags design.
3. No assertion that `scheduled_downgrade` is reset to `False` on the create path (a cheap regression guard).

---

## 7. Summary of Recommendations

| # | Severity | Recommendation | Blocker? |
|---|----------|----------------|----------|
| §3 | Medium | Confirm the upload/premium runtime path lazily ensures the S3 bucket; if not, provision it here. Note the partial-provisioning boundary in the PR. | No, but important |
| §4.1 | Medium | Add a test for the `_insert_or_reselect` `IntegrityError` → re-select fallback (and the `None` re-raise). | No, strongly recommended |
| §4.2 | Medium | Emit a `WARNING` log when the admin path creates (not updates) a row, so systemic row loss is not masked. | No |
| §4.3 | Low–Med | Correct the "confined to the test module" claim for `@compiles`, or scope the SQLite BIGINT coercion to a dedicated engine. | No |
| §4.4 | Low | Confirm no frontend flow relied on premium silent-coercion to `now()`. | No |
| §3 / §4.5 | Low | Use the canonical `StorageQuota.FREE` constant when creating a free-plan storage row; confirm a scan reconciles `storage_usage_bytes=0`. | No |
| §4.6 | Nit | Type hints on `_insert_or_reselect`; add inverse mixed-state + `scheduled_downgrade` assertions. | No |

**Net:** The design is sound and the implementation is careful; the dynamic creation mirrors existing signup provisioning for the two rows it touches and is safe per row. The work left is to (a) close the partial-provisioning gap (bucket + canonical free quota) or document its bounds, (b) test the concurrency guard, and (c) surface an operational signal so silent repair does not hide a systemic cause.

**Review by itutu-tienday — APPROVED** [review]
**Verdict: all code-actionable items are resolved and the fixes are correct — approving.** Three "please confirm" items remain (non-blocking).

## Resolved

| Finding | Status | Notes |
|---------|--------|-------|
| **4.1** — concurrency guard untested | ✅ Resolved | `TestInsertOrReselect` covers both branches: unique conflict → returns the existing winner, not the duplicate (asserts `count == 1`); NOT-NULL `IntegrityError` with no row to re-select → re-raises (exercises `if existing is None: raise`). Semantics check out: `begin_nested()` rolls back only the SAVEPOINT, so the outer transaction survives and the re-select runs cleanly; on unique conflict the rolled-back `duplicate` is expunged, keeping the count at 1. |
| **4.2** — no signal on implicit repair | ✅ Resolved | The `WARNING` names exactly which of `subscription_users` / `user_storage_usage` was created. Nicely tested three ways: both-missing (names both), update path (no warning), and only-storage-missing (names storage, not subscription — this last one also guards against an inverted label/flag pairing). |
| **4.3** — `@compiles` mislabeled as module-scoped | ✅ Resolved | Comment corrected to state it mutates SQLAlchemy's process-global dialect registry on import. Documenting rather than scoping to a dedicated engine is a fine call given the options; the behavior is now honestly described. |
| **4.6** — inverse mixed state, `scheduled_downgrade`, type hints | ✅ Resolved | Added `test_creates_storage_when_only_subscription_exists`, a `scheduled_downgrade is False` assertion on the create path, and typed `_insert_or_reselect` via `_UniqueRow = TypeVar` with the expected models named in the docstring. |

## Remaining — confirm only, non-blocking

1. **Bucket assumption (§3).** The upsert restores the two rows but not the S3 bucket, so a *genuinely never-provisioned* user could reach an upload/premium flow without one. Could you confirm the upload / premium-activation runtime path lazily calls `ensure_user_bucket_exists`? If it does, this is fully benign — a one-line note in the PR would close it out. Fine as a follow-up either way.
2. **Premium-without-expiration is now a 400 on the update path too (§4.4).** Correct fix (it prevented a born-expired row), but please confirm no frontend flow relied on the old silent `expiration = now()` coercion.
3. **Free-plan quota (§4.5, low).** Creating a free-plan row still stores the admin-submitted quota rather than the canonical `StorageQuota.FREE` constant. Admin discretion — optional.

None of these block merge. Once the three confirms above are acknowledged, this is good to go. 👍
